{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "from scipy.stats import entropy\n",
    "from scipy.stats import mannwhitneyu as mwu\n",
    "import plotly.graph_objects as go\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################\n",
    "### TCR, BCR\n",
    "dir_BCR_01H = '../Data/BulkBCRseq/1st/BCR_HC_CDR3/'\n",
    "dir_BCR_01L = '../Data/BulkBCRseq/1st/BCR_LC_CDR3/'\n",
    "dir_BCR_02H = '../Data/BulkBCRseq/2nd/BCR_HC_CDR3/'\n",
    "dir_BCR_02L = '../Data/BulkBCRseq/2nd/BCR_LC_CDR3/'\n",
    "dir_TCR_001 = '../Data/BulkTCRseq/1st/'\n",
    "dir_TCR_002 = '../Data/BulkTCRseq/2nd/'\n",
    "###################################################\n",
    "### COVIDseq, HLA, cytokine\n",
    "dir_COV_01 = '../Data/COVIDseq/1st/whole/'\n",
    "dir_COV_02 = '../Data/COVIDseq/2nd/'\n",
    "dir_HLA_01 = '../Data/HLA_typing/1st/'\n",
    "dir_HLA_02 = '../Data/HLA_typing/2nd/'\n",
    "dir_CYT_01 = '../Data/Cytokine_Profiling/1st/'\n",
    "dir_CYT_02 = '../Data/Cytokine_Profiling/2nd/'\n",
    "\n",
    "###################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def short_view(directory):\n",
    "    all_dfs = []\n",
    "    filenames = os.listdir(directory) \n",
    "    if filenames:\n",
    "        filename = filenames[10]\n",
    "        if filename.startswith('COV') and filename.endswith('.csv') or filename.endswith('.txt'):\n",
    "            df = pd.read_csv(directory + filename, sep=',')\n",
    "            df['Directory'] = directory.split('/')[-2]\n",
    "            df.insert(0, 'Directory', df.pop('Directory'))\n",
    "            all_dfs.append(df.head(n=3))\n",
    "            \n",
    "    result = pd.concat(all_dfs)\n",
    "    display(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Directory</th>\n",
       "      <th>Read Count</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>Clonal Sequence</th>\n",
       "      <th>Clonal Sequence Quality</th>\n",
       "      <th>CDR3 Min Quality</th>\n",
       "      <th>CDR3 Sequence</th>\n",
       "      <th>CDR3 Amino Acid Sequence</th>\n",
       "      <th>Clonal Type</th>\n",
       "      <th>Frame Shift</th>\n",
       "      <th>Stop Codon</th>\n",
       "      <th>Amino Acid Length</th>\n",
       "      <th>V segment</th>\n",
       "      <th>all V hits</th>\n",
       "      <th>D segment</th>\n",
       "      <th>all D hits</th>\n",
       "      <th>J segment</th>\n",
       "      <th>all J hits</th>\n",
       "      <th>C segment</th>\n",
       "      <th>all C hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BCR_HC_CDR3</td>\n",
       "      <td>2459</td>\n",
       "      <td>0.068749</td>\n",
       "      <td>TGTGCAAAAGGAGGCCGTATATCAGCAGATGGTCCTTCTGATTCTTGG</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCAAAAGGAGGCCGTATATCAGCAGATGGTCCTTCTGATTCTTGG</td>\n",
       "      <td>CAKGGRISADGPSDSW</td>\n",
       "      <td>IGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>IGHV3-23</td>\n",
       "      <td>IGHD6-13</td>\n",
       "      <td>IGHD6-13</td>\n",
       "      <td>IGHJ1</td>\n",
       "      <td>IGHJ1;IGHJ4;IGHJ5</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>IGHG1;IGHG2;IGHGP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR_HC_CDR3</td>\n",
       "      <td>1974</td>\n",
       "      <td>0.055189</td>\n",
       "      <td>TGTGCGAGAGGCGCAAGAACTATACTTCGGGCCTGG</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCGAGAGGCGCAAGAACTATACTTCGGGCCTGG</td>\n",
       "      <td>CARGARTILRAW</td>\n",
       "      <td>IGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>IGHV4-34</td>\n",
       "      <td>IGHV4-34</td>\n",
       "      <td>IGHD3-10</td>\n",
       "      <td>IGHD3-10;IGHD1-26;IGHD6-6</td>\n",
       "      <td>IGHJ5</td>\n",
       "      <td>IGHJ5;IGHJ4</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>IGHG1;IGHG2;IGHGP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR_HC_CDR3</td>\n",
       "      <td>1970</td>\n",
       "      <td>0.055077</td>\n",
       "      <td>TGTGCGAGGGGGAAACGTTCCATTACTATGATACTAGTGGTTATAG...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCGAGGGGGAAACGTTCCATTACTATGATACTAGTGGTTATAG...</td>\n",
       "      <td>CARGKRSITMILVVIGTEYYFDYW</td>\n",
       "      <td>IGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "      <td>IGHV4-34</td>\n",
       "      <td>IGHV4-34</td>\n",
       "      <td>IGHD3-22</td>\n",
       "      <td>IGHD3-22</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHJ4</td>\n",
       "      <td>IGHG1</td>\n",
       "      <td>IGHG1;IGHG2;IGHGP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Directory  Read Count  Fraction  \\\n",
       "0  BCR_HC_CDR3        2459  0.068749   \n",
       "1  BCR_HC_CDR3        1974  0.055189   \n",
       "2  BCR_HC_CDR3        1970  0.055077   \n",
       "\n",
       "                                     Clonal Sequence  \\\n",
       "0   TGTGCAAAAGGAGGCCGTATATCAGCAGATGGTCCTTCTGATTCTTGG   \n",
       "1               TGTGCGAGAGGCGCAAGAACTATACTTCGGGCCTGG   \n",
       "2  TGTGCGAGGGGGAAACGTTCCATTACTATGATACTAGTGGTTATAG...   \n",
       "\n",
       "                             Clonal Sequence Quality  CDR3 Min Quality  \\\n",
       "0   IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "1               IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "2  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII...                40   \n",
       "\n",
       "                                       CDR3 Sequence  \\\n",
       "0   TGTGCAAAAGGAGGCCGTATATCAGCAGATGGTCCTTCTGATTCTTGG   \n",
       "1               TGTGCGAGAGGCGCAAGAACTATACTTCGGGCCTGG   \n",
       "2  TGTGCGAGGGGGAAACGTTCCATTACTATGATACTAGTGGTTATAG...   \n",
       "\n",
       "   CDR3 Amino Acid Sequence Clonal Type Frame Shift Stop Codon  \\\n",
       "0          CAKGGRISADGPSDSW         IGG         NaN        NaN   \n",
       "1              CARGARTILRAW         IGG         NaN        NaN   \n",
       "2  CARGKRSITMILVVIGTEYYFDYW         IGG         NaN        NaN   \n",
       "\n",
       "   Amino Acid Length V segment all V hits D segment  \\\n",
       "0                 16  IGHV3-23   IGHV3-23  IGHD6-13   \n",
       "1                 12  IGHV4-34   IGHV4-34  IGHD3-10   \n",
       "2                 24  IGHV4-34   IGHV4-34  IGHD3-22   \n",
       "\n",
       "                  all D hits J segment         all J hits C segment  \\\n",
       "0                   IGHD6-13     IGHJ1  IGHJ1;IGHJ4;IGHJ5     IGHG1   \n",
       "1  IGHD3-10;IGHD1-26;IGHD6-6     IGHJ5        IGHJ5;IGHJ4     IGHG1   \n",
       "2                   IGHD3-22     IGHJ4              IGHJ4     IGHG1   \n",
       "\n",
       "          all C hits  \n",
       "0  IGHG1;IGHG2;IGHGP  \n",
       "1  IGHG1;IGHG2;IGHGP  \n",
       "2  IGHG1;IGHG2;IGHGP  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Directory</th>\n",
       "      <th>Read Count</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>Clonal Sequence</th>\n",
       "      <th>Clonal Sequence Quality</th>\n",
       "      <th>CDR3 Min Quality</th>\n",
       "      <th>CDR3 Sequence</th>\n",
       "      <th>CDR3 Amino Acid Sequence</th>\n",
       "      <th>Clonal Type</th>\n",
       "      <th>Frame Shift</th>\n",
       "      <th>Stop Codon</th>\n",
       "      <th>Amino Acid Length</th>\n",
       "      <th>V segment</th>\n",
       "      <th>all V hits</th>\n",
       "      <th>D segment</th>\n",
       "      <th>all D hits</th>\n",
       "      <th>J segment</th>\n",
       "      <th>all J hits</th>\n",
       "      <th>C segment</th>\n",
       "      <th>all C hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BCR_LC_CDR3</td>\n",
       "      <td>8270</td>\n",
       "      <td>0.081901</td>\n",
       "      <td>TGTCAGCAATATTATCTTAGTCCCTTCACTTTT</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTCAGCAATATTATCTTAGTCCCTTCACTTTT</td>\n",
       "      <td>CQQYYLSPFTF</td>\n",
       "      <td>IGK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGKJ2</td>\n",
       "      <td>IGKJ2</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>IGKC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR_LC_CDR3</td>\n",
       "      <td>2522</td>\n",
       "      <td>0.024976</td>\n",
       "      <td>TGTCAACAACTTAGTAGTTACCCCCTCACTTTC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTCAACAACTTAGTAGTTACCCCCTCACTTTC</td>\n",
       "      <td>CQQLSSYPLTF</td>\n",
       "      <td>IGK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>IGKV1-9</td>\n",
       "      <td>IGKV1-9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGKJ4</td>\n",
       "      <td>IGKJ4</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>IGKC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR_LC_CDR3</td>\n",
       "      <td>2479</td>\n",
       "      <td>0.024551</td>\n",
       "      <td>TGTCAGCAATATTATAGTACTCCTCGGACGTTC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTCAGCAATATTATAGTACTCCTCGGACGTTC</td>\n",
       "      <td>CQQYYSTPRTF</td>\n",
       "      <td>IGK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>IGKV4-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGKJ1</td>\n",
       "      <td>IGKJ1</td>\n",
       "      <td>IGKC</td>\n",
       "      <td>IGKC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Directory  Read Count  Fraction                    Clonal Sequence  \\\n",
       "0  BCR_LC_CDR3        8270  0.081901  TGTCAGCAATATTATCTTAGTCCCTTCACTTTT   \n",
       "1  BCR_LC_CDR3        2522  0.024976  TGTCAACAACTTAGTAGTTACCCCCTCACTTTC   \n",
       "2  BCR_LC_CDR3        2479  0.024551  TGTCAGCAATATTATAGTACTCCTCGGACGTTC   \n",
       "\n",
       "             Clonal Sequence Quality  CDR3 Min Quality  \\\n",
       "0  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "1  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "2  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "\n",
       "                       CDR3 Sequence CDR3 Amino Acid Sequence Clonal Type  \\\n",
       "0  TGTCAGCAATATTATCTTAGTCCCTTCACTTTT              CQQYYLSPFTF         IGK   \n",
       "1  TGTCAACAACTTAGTAGTTACCCCCTCACTTTC              CQQLSSYPLTF         IGK   \n",
       "2  TGTCAGCAATATTATAGTACTCCTCGGACGTTC              CQQYYSTPRTF         IGK   \n",
       "\n",
       "  Frame Shift Stop Codon  Amino Acid Length V segment all V hits  D segment  \\\n",
       "0         NaN        NaN                 11   IGKV4-1    IGKV4-1        NaN   \n",
       "1         NaN        NaN                 11   IGKV1-9    IGKV1-9        NaN   \n",
       "2         NaN        NaN                 11   IGKV4-1    IGKV4-1        NaN   \n",
       "\n",
       "   all D hits J segment all J hits C segment all C hits  \n",
       "0         NaN     IGKJ2      IGKJ2      IGKC       IGKC  \n",
       "1         NaN     IGKJ4      IGKJ4      IGKC       IGKC  \n",
       "2         NaN     IGKJ1      IGKJ1      IGKC       IGKC  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Directory</th>\n",
       "      <th>Read Count</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>Clonal Sequence</th>\n",
       "      <th>Clonal Sequence Quality</th>\n",
       "      <th>CDR3 Min Quality</th>\n",
       "      <th>CDR3 Sequence</th>\n",
       "      <th>CDR3 Amino Acid Sequence</th>\n",
       "      <th>Clonal Type</th>\n",
       "      <th>Frame Shift</th>\n",
       "      <th>Stop Codon</th>\n",
       "      <th>Amino Acid Length</th>\n",
       "      <th>V segment</th>\n",
       "      <th>all V hits</th>\n",
       "      <th>D segment</th>\n",
       "      <th>all D hits</th>\n",
       "      <th>J segment</th>\n",
       "      <th>all J hits</th>\n",
       "      <th>C segment</th>\n",
       "      <th>all C hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Directory, Read Count, Fraction, Clonal Sequence, Clonal Sequence Quality, CDR3 Min Quality, CDR3 Sequence, CDR3 Amino Acid Sequence, Clonal Type, Frame Shift, Stop Codon, Amino Acid Length, V segment, all V hits, D segment, all D hits, J segment, all J hits, C segment, all C hits]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Directory</th>\n",
       "      <th>Read Count</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>Clonal Sequence</th>\n",
       "      <th>Clonal Sequence Quality</th>\n",
       "      <th>CDR3 Min Quality</th>\n",
       "      <th>CDR3 Sequence</th>\n",
       "      <th>CDR3 Amino Acid Sequence</th>\n",
       "      <th>Clonal Type</th>\n",
       "      <th>Frame Shift</th>\n",
       "      <th>Stop Codon</th>\n",
       "      <th>Amino Acid Length</th>\n",
       "      <th>V segment</th>\n",
       "      <th>all V hits</th>\n",
       "      <th>D segment</th>\n",
       "      <th>all D hits</th>\n",
       "      <th>J segment</th>\n",
       "      <th>all J hits</th>\n",
       "      <th>C segment</th>\n",
       "      <th>all C hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BCR_LC_CDR3</td>\n",
       "      <td>651</td>\n",
       "      <td>0.053453</td>\n",
       "      <td>TGCTGCTCATATGCAGGTAGTAGCCCCGTGGTATTC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGCTGCTCATATGCAGGTAGTAGCCCCGTGGTATTC</td>\n",
       "      <td>CCSYAGSSPVVF</td>\n",
       "      <td>IGL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>IGLV2-23</td>\n",
       "      <td>IGLV2-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>IGLC7</td>\n",
       "      <td>IGLC7;IGLC2;IGLC3;IGLC6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR_LC_CDR3</td>\n",
       "      <td>460</td>\n",
       "      <td>0.037770</td>\n",
       "      <td>TGTCAGACCTGGGGCACTGGCATTCAGGTATTC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTCAGACCTGGGGCACTGGCATTCAGGTATTC</td>\n",
       "      <td>CQTWGTGIQVF</td>\n",
       "      <td>IGL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>IGLV4-69</td>\n",
       "      <td>IGLV4-69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>IGLJ2;IGLJ3</td>\n",
       "      <td>IGLC7</td>\n",
       "      <td>IGLC7;IGLC2;IGLC3;IGLC6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCR_LC_CDR3</td>\n",
       "      <td>352</td>\n",
       "      <td>0.028902</td>\n",
       "      <td>TGCTGCTCATATGCAGCTAGTAGCCCCGTGGTATTC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGCTGCTCATATGCAGCTAGTAGCCCCGTGGTATTC</td>\n",
       "      <td>CCSYAASSPVVF</td>\n",
       "      <td>IGL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>IGLV2-23</td>\n",
       "      <td>IGLV2-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>IGLJ2</td>\n",
       "      <td>IGLC7</td>\n",
       "      <td>IGLC7;IGLC2;IGLC3;IGLC6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Directory  Read Count  Fraction                       Clonal Sequence  \\\n",
       "0  BCR_LC_CDR3         651  0.053453  TGCTGCTCATATGCAGGTAGTAGCCCCGTGGTATTC   \n",
       "1  BCR_LC_CDR3         460  0.037770     TGTCAGACCTGGGGCACTGGCATTCAGGTATTC   \n",
       "2  BCR_LC_CDR3         352  0.028902  TGCTGCTCATATGCAGCTAGTAGCCCCGTGGTATTC   \n",
       "\n",
       "                Clonal Sequence Quality  CDR3 Min Quality  \\\n",
       "0  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "1     IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "2  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "\n",
       "                          CDR3 Sequence CDR3 Amino Acid Sequence Clonal Type  \\\n",
       "0  TGCTGCTCATATGCAGGTAGTAGCCCCGTGGTATTC             CCSYAGSSPVVF         IGL   \n",
       "1     TGTCAGACCTGGGGCACTGGCATTCAGGTATTC              CQTWGTGIQVF         IGL   \n",
       "2  TGCTGCTCATATGCAGCTAGTAGCCCCGTGGTATTC             CCSYAASSPVVF         IGL   \n",
       "\n",
       "  Frame Shift Stop Codon  Amino Acid Length V segment all V hits  D segment  \\\n",
       "0         NaN        NaN                 12  IGLV2-23   IGLV2-23        NaN   \n",
       "1         NaN        NaN                 11  IGLV4-69   IGLV4-69        NaN   \n",
       "2         NaN        NaN                 12  IGLV2-23   IGLV2-23        NaN   \n",
       "\n",
       "   all D hits J segment   all J hits C segment               all C hits  \n",
       "0         NaN     IGLJ2        IGLJ2     IGLC7  IGLC7;IGLC2;IGLC3;IGLC6  \n",
       "1         NaN     IGLJ2  IGLJ2;IGLJ3     IGLC7  IGLC7;IGLC2;IGLC3;IGLC6  \n",
       "2         NaN     IGLJ2        IGLJ2     IGLC7  IGLC7;IGLC2;IGLC3;IGLC6  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Directory</th>\n",
       "      <th>Read Count</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>Clonal Sequence</th>\n",
       "      <th>Clonal Sequence Quality</th>\n",
       "      <th>CDR3 Min Quality</th>\n",
       "      <th>CDR3 Sequence</th>\n",
       "      <th>CDR3 Amino Acid Sequence</th>\n",
       "      <th>Clonal Type</th>\n",
       "      <th>Frame Shift</th>\n",
       "      <th>Stop Codon</th>\n",
       "      <th>Amino Acid Length</th>\n",
       "      <th>V segment</th>\n",
       "      <th>all V hits</th>\n",
       "      <th>D segment</th>\n",
       "      <th>all D hits</th>\n",
       "      <th>J segment</th>\n",
       "      <th>all J hits</th>\n",
       "      <th>C segment</th>\n",
       "      <th>all C hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1st</td>\n",
       "      <td>243</td>\n",
       "      <td>0.012202</td>\n",
       "      <td>TGCATCCTGAGAGACGACAATGACATGCGCTTT</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGCATCCTGAGAGACGACAATGACATGCGCTTT</td>\n",
       "      <td>CILRDDNDMRF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>TRAV26-2</td>\n",
       "      <td>TRAV26-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ43</td>\n",
       "      <td>TRAJ43</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>TRAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1st</td>\n",
       "      <td>141</td>\n",
       "      <td>0.007080</td>\n",
       "      <td>TGTGCTCTGAGTGAGGCGGGGGACACGGGCAGGAGAGCACTTACTTTT</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCTCTGAGTGAGGCGGGGGACACGGGCAGGAGAGCACTTACTTTT</td>\n",
       "      <td>CALSEAGDTGRRALTF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16</td>\n",
       "      <td>TRAV19</td>\n",
       "      <td>TRAV19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ5</td>\n",
       "      <td>TRAJ5</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>TRAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1st</td>\n",
       "      <td>93</td>\n",
       "      <td>0.004670</td>\n",
       "      <td>TGTGCTGTGAGAGACATAGTGGATAGCAACTATCAGTTAATCTGG</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCTGTGAGAGACATAGTGGATAGCAACTATCAGTTAATCTGG</td>\n",
       "      <td>CAVRDIVDSNYQLIW</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>TRAV3</td>\n",
       "      <td>TRAV3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ33</td>\n",
       "      <td>TRAJ33</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>TRAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Directory  Read Count  Fraction  \\\n",
       "0       1st         243  0.012202   \n",
       "1       1st         141  0.007080   \n",
       "2       1st          93  0.004670   \n",
       "\n",
       "                                    Clonal Sequence  \\\n",
       "0                 TGCATCCTGAGAGACGACAATGACATGCGCTTT   \n",
       "1  TGTGCTCTGAGTGAGGCGGGGGACACGGGCAGGAGAGCACTTACTTTT   \n",
       "2     TGTGCTGTGAGAGACATAGTGGATAGCAACTATCAGTTAATCTGG   \n",
       "\n",
       "                            Clonal Sequence Quality  CDR3 Min Quality  \\\n",
       "0                 IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "1  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "2     IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "\n",
       "                                      CDR3 Sequence CDR3 Amino Acid Sequence  \\\n",
       "0                 TGCATCCTGAGAGACGACAATGACATGCGCTTT              CILRDDNDMRF   \n",
       "1  TGTGCTCTGAGTGAGGCGGGGGACACGGGCAGGAGAGCACTTACTTTT         CALSEAGDTGRRALTF   \n",
       "2     TGTGCTGTGAGAGACATAGTGGATAGCAACTATCAGTTAATCTGG          CAVRDIVDSNYQLIW   \n",
       "\n",
       "  Clonal Type Frame Shift Stop Codon  Amino Acid Length V segment all V hits  \\\n",
       "0         TRA         NaN        NaN                 11  TRAV26-2   TRAV26-2   \n",
       "1         TRA         NaN        NaN                 16    TRAV19     TRAV19   \n",
       "2         TRA         NaN        NaN                 15     TRAV3      TRAV3   \n",
       "\n",
       "   D segment  all D hits J segment all J hits C segment all C hits  \n",
       "0        NaN         NaN    TRAJ43     TRAJ43      TRAC       TRAC  \n",
       "1        NaN         NaN     TRAJ5      TRAJ5      TRAC       TRAC  \n",
       "2        NaN         NaN    TRAJ33     TRAJ33      TRAC       TRAC  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Directory</th>\n",
       "      <th>Read Count</th>\n",
       "      <th>Fraction</th>\n",
       "      <th>Clonal Sequence</th>\n",
       "      <th>Clonal Sequence Quality</th>\n",
       "      <th>CDR3 Min Quality</th>\n",
       "      <th>CDR3 Sequence</th>\n",
       "      <th>CDR3 Amino Acid Sequence</th>\n",
       "      <th>Clonal Type</th>\n",
       "      <th>Frame Shift</th>\n",
       "      <th>Stop Codon</th>\n",
       "      <th>Amino Acid Length</th>\n",
       "      <th>V segment</th>\n",
       "      <th>all V hits</th>\n",
       "      <th>D segment</th>\n",
       "      <th>all D hits</th>\n",
       "      <th>J segment</th>\n",
       "      <th>all J hits</th>\n",
       "      <th>C segment</th>\n",
       "      <th>all C hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2nd</td>\n",
       "      <td>132</td>\n",
       "      <td>0.011495</td>\n",
       "      <td>TGTGCTACGGACGCGGGGGATGACAAGATCATCTTT</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCTACGGACGCGGGGGATGACAAGATCATCTTT</td>\n",
       "      <td>CATDAGDDKIIF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>TRAV17</td>\n",
       "      <td>TRAV17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ30</td>\n",
       "      <td>TRAJ30</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>TRAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2nd</td>\n",
       "      <td>130</td>\n",
       "      <td>0.011321</td>\n",
       "      <td>TGTGCAGCATTGAGAGATGGAGGAAGCCAAGGAAATCTCATCTTT</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCAGCATTGAGAGATGGAGGAAGCCAAGGAAATCTCATCTTT</td>\n",
       "      <td>CAALRDGGSQGNLIF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>TRAV29DV5</td>\n",
       "      <td>TRAV29DV5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>TRAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2nd</td>\n",
       "      <td>122</td>\n",
       "      <td>0.010624</td>\n",
       "      <td>TGTGCTGTGAGAGACTTGGACATGGGCAGGAGAGCACTTACTTTT</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII</td>\n",
       "      <td>40</td>\n",
       "      <td>TGTGCTGTGAGAGACTTGGACATGGGCAGGAGAGCACTTACTTTT</td>\n",
       "      <td>CAVRDLDMGRRALTF</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15</td>\n",
       "      <td>TRAV3</td>\n",
       "      <td>TRAV3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRAJ5</td>\n",
       "      <td>TRAJ5</td>\n",
       "      <td>TRAC</td>\n",
       "      <td>TRAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Directory  Read Count  Fraction  \\\n",
       "0       2nd         132  0.011495   \n",
       "1       2nd         130  0.011321   \n",
       "2       2nd         122  0.010624   \n",
       "\n",
       "                                 Clonal Sequence  \\\n",
       "0           TGTGCTACGGACGCGGGGGATGACAAGATCATCTTT   \n",
       "1  TGTGCAGCATTGAGAGATGGAGGAAGCCAAGGAAATCTCATCTTT   \n",
       "2  TGTGCTGTGAGAGACTTGGACATGGGCAGGAGAGCACTTACTTTT   \n",
       "\n",
       "                         Clonal Sequence Quality  CDR3 Min Quality  \\\n",
       "0           IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "1  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "2  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII                40   \n",
       "\n",
       "                                   CDR3 Sequence CDR3 Amino Acid Sequence  \\\n",
       "0           TGTGCTACGGACGCGGGGGATGACAAGATCATCTTT             CATDAGDDKIIF   \n",
       "1  TGTGCAGCATTGAGAGATGGAGGAAGCCAAGGAAATCTCATCTTT          CAALRDGGSQGNLIF   \n",
       "2  TGTGCTGTGAGAGACTTGGACATGGGCAGGAGAGCACTTACTTTT          CAVRDLDMGRRALTF   \n",
       "\n",
       "  Clonal Type Frame Shift Stop Codon  Amino Acid Length  V segment all V hits  \\\n",
       "0         TRA         NaN        NaN                 12     TRAV17     TRAV17   \n",
       "1         TRA         NaN        NaN                 15  TRAV29DV5  TRAV29DV5   \n",
       "2         TRA         NaN        NaN                 15      TRAV3      TRAV3   \n",
       "\n",
       "   D segment  all D hits J segment all J hits C segment all C hits  \n",
       "0        NaN         NaN    TRAJ30     TRAJ30      TRAC       TRAC  \n",
       "1        NaN         NaN    TRAJ42     TRAJ42      TRAC       TRAC  \n",
       "2        NaN         NaN     TRAJ5      TRAJ5      TRAC       TRAC  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dir_list = [dir_BCR_01H, dir_BCR_01L, dir_BCR_02H, dir_BCR_02L, dir_TCR_001, dir_TCR_002]\n",
    "for i in dir_list:\n",
    "    short_view(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>PNTTM</th>\n",
       "      <th>COVID19</th>\n",
       "      <th>NPRONG</th>\n",
       "      <th>MASK</th>\n",
       "      <th>HFNC</th>\n",
       "      <th>NIV</th>\n",
       "      <th>IV</th>\n",
       "      <th>ECMO</th>\n",
       "      <th>CRRT</th>\n",
       "      <th>NOGAS</th>\n",
       "      <th>XRAYT</th>\n",
       "      <th>ER</th>\n",
       "      <th>severity</th>\n",
       "      <th>severity_init</th>\n",
       "      <th>from</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>484</td>\n",
       "      <td>COV-CCO-041</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>492</td>\n",
       "      <td>COV-CCO-041</td>\n",
       "      <td>9</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>493</td>\n",
       "      <td>COV-CCO-041</td>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0           ID  PNTTM  COVID19  NPRONG  MASK  HFNC  NIV   IV  \\\n",
       "0         484  COV-CCO-041      1      1.0     2.0   2.0   2.0  2.0  2.0   \n",
       "1         492  COV-CCO-041      9      1.0     2.0   2.0   2.0  2.0  2.0   \n",
       "2         493  COV-CCO-041     10      1.0     2.0   2.0   2.0  2.0  2.0   \n",
       "\n",
       "   ECMO  CRRT  NOGAS  XRAYT   ER  severity  severity_init  from  \n",
       "0   2.0   2.0    2.0    2.0  1.0         1            1.0     1  \n",
       "1   2.0   2.0    2.0    2.0  1.0         1            1.0     1  \n",
       "2   2.0   2.0    2.0    2.0  1.0         1            1.0     1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## get max_css\n",
    "\n",
    "data_dir = '../Result/useful_table/'\n",
    "max_css = pd.read_csv(data_dir + \"change_css_change_only_adaptome.csv\")\\\n",
    "\n",
    "display(max_css.head(n=3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "265\n",
      "85\n",
      "['COV-CNO-029', 'COV-CNO-030', 'COV-CNO-032', 'COV-CNO-033', 'COV-CNO-034', 'COV-CNO-035', 'COV-CNO-036', 'COV-CNO-037', 'COV-CNO-038', 'COV-CNO-039', 'COV-CNO-040', 'COV-CNO-041', 'COV-CNO-042', 'COV-CNO-043', 'COV-CNO-044', 'COV-CNO-045', 'COV-CNO-046', 'COV-CNO-048', 'COV-CNO-049', 'COV-CNO-050', 'COV-CNO-051', 'COV-CNO-052', 'COV-CNO-053', 'COV-CNO-054', 'COV-CNO-055', 'COV-CNO-056', 'COV-CNO-057', 'COV-CNO-058', 'COV-CNO-059', 'COV-CNO-060', 'COV-CNO-061', 'COV-CNO-065', 'COV-CNO-066', 'COV-CNO-067', 'COV-CNO-068', 'COV-CNO-069', 'COV-CNO-070', 'COV-CNO-071', 'COV-CNO-073', 'COV-CNO-080', 'COV-CCO-239', 'COV-CCO-240', 'COV-CCO-241', 'COV-CCO-258', 'COV-CNC-001', 'COV-CNC-002', 'COV-CNC-004', 'COV-CNC-005', 'COV-CNC-010', 'COV-CNC-013', 'COV-CNC-014', 'COV-CNC-015', 'COV-CNC-019', 'COV-CNC-020', 'COV-CNC-022', 'COV-CNC-023', 'COV-CNC-025', 'COV-CNC-030', 'COV-CNC-032', 'COV-CNC-035', 'COV-CNC-036', 'COV-CNC-039', 'COV-CNC-040', 'COV-CNC-041', 'COV-CNC-042', 'COV-CNC-045', 'COV-CNC-046', 'COV-CNC-047', 'COV-CNC-049', 'COV-CNC-052', 'COV-CNC-053', 'COV-CNC-055', 'COV-CNC-056', 'COV-CNC-057', 'COV-CNC-058', 'COV-CNC-060', 'COV-CNC-101', 'COV-CNC-103', 'COV-CNC-105', 'COV-CNC-107', 'COV-CNC-109', 'COV-CNC-111', 'COV-CNC-113', 'COV-CNC-115', 'COV-CNC-117']\n"
     ]
    }
   ],
   "source": [
    "cov_list = [dir_COV_01, dir_COV_02]\n",
    "\n",
    "### list 내의 모든 파일들을 읽어서 파일명 앞 11자리를 name_list에 저장\n",
    "\n",
    "def get_name_list(directory):\n",
    "    name_list = []\n",
    "    filenames = os.listdir(directory)\n",
    "    for filename in filenames:\n",
    "        if filename.startswith('COV') and filename.endswith('.csv') or filename.endswith('.txt'):\n",
    "            name_list.append(filename[:11])\n",
    "    return name_list\n",
    "\n",
    "total_name_list = [] \n",
    " \n",
    "for i in cov_list:\n",
    "    total_name_list += get_name_list(i)\n",
    "    \n",
    "### total_name_list에 있는 ID가 max_css에 있는지 확인\n",
    "\n",
    "### 중복 ID제외하고\n",
    "max_list = max_css['ID'].unique()\n",
    "print(len(max_list))\n",
    "\n",
    "not_in_list = []\n",
    "not_adaptome = []\n",
    "\n",
    "for i in total_name_list:\n",
    "    if i not in max_list:\n",
    "        not_in_list.append(i)\n",
    "\n",
    "for j in max_list:\n",
    "    if j not in total_name_list:\n",
    "        not_adaptome.append(j)\n",
    "\n",
    "### not_in_list에 있는 ID의 unique한 개수 확인\n",
    "\n",
    "\n",
    "print(len(not_adaptome))\n",
    "print(not_adaptome)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "620\n",
      "555\n",
      "0\n",
      "4\n",
      "['COV-CNC-047', 'COV-CNC-057', 'COV-CNC-058', 'COV-CNC-060']\n"
     ]
    }
   ],
   "source": [
    "hla_list = [dir_HLA_01, dir_HLA_02]\n",
    "cyt_list = [dir_CYT_01, dir_CYT_02]\n",
    "\n",
    "hla_name_list = []\n",
    "cyt_name_list = []\n",
    "\n",
    "for i in hla_list:\n",
    "    hla_name_list += get_name_list(i)\n",
    "for i in cyt_list:\n",
    "    cyt_name_list += get_name_list(i)\n",
    "    \n",
    "# hla_name_list = list(set(hla_name_list))\n",
    "print(len(hla_name_list))\n",
    "\n",
    "### 중복 제외하고 cyt_name_list 개수    \n",
    "cyt_name_list = list(set(cyt_name_list))\n",
    "print(len(cyt_name_list))\n",
    "\n",
    "not_hla = []\n",
    "not_cyt = []\n",
    "\n",
    "for i in max_list:\n",
    "    if i not in hla_name_list:\n",
    "        not_hla.append(i)\n",
    "    if i not in cyt_name_list:\n",
    "        not_cyt.append(i)\n",
    "        \n",
    "print(len(not_hla))\n",
    "print(len(not_cyt))\n",
    "print(not_cyt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ID', 'alpha.2.Macroglobulin.pg.ml.', 'beta.2.Microglobulin.pg.ml.',\n",
      "       'Fetuin.A.AHSG.pg.ml.', 'MBL.pg.ml.', 'Protein.S.PROS1.pg.ml.',\n",
      "       'Serpin.A4.Kallistatin.pg.ml.', 'CD14.pg.ml.', 'Component.C2.pg.ml.',\n",
      "       'CXCL4.PF4.pg.ml.',\n",
      "       ...\n",
      "       'u.Plasminogen.Activator.uPA.Urokinase.pg.ml.', 'CA15.3.MUC.1.pg.ml.',\n",
      "       'Complement.Component.C9.pg.ml.', 'TGF.beta1.pg.ml.',\n",
      "       'Aldehyde.Dehydrogenase.1.A1.ALDH1A1.pg.ml.',\n",
      "       'CCL14.HCC.1.HCC.3.pg.ml.', 'Complement.Component.C2.pg.ml.',\n",
      "       'Serpin.C1.Antithrombin.III.pg.ml.', 'beta2.Microglobulin.pg.ml.',\n",
      "       'CXCL12.pg.ml.'],\n",
      "      dtype='object', length=227)\n"
     ]
    }
   ],
   "source": [
    "### cyt_list의 경로에 있는 모든 파일에 대해 \n",
    "# Sample.ID.\talpha.2.Macroglobulin..pg.ml.\tbeta.2.Microglobulin..pg.ml.\tFetuin.A.AHSG..pg.ml.\tMBL..pg.ml.\tProtein.S.PROS1..pg.ml.\tSerpin.A4.Kallistatin..pg.ml.\tCD14..pg.ml.\tComponent.C2..pg.ml.\tCXCL4.PF4..pg.ml.\tLBP..pg.ml.\tLRG1..pg.ml.\tProperdin..pg.ml.\tSerpin.C1..pg.ml.\tSerpin.A10.ZPI..pg.ml.\tAldehyd.Dehydrogenase.1.A1..pg.ml.\tBCMA.TNFRSF17..pg.ml.\tCCL14..pg.ml.\tCCL18.PARC..pg.ml.\tCoagulation.Factor.XIV.Protein.C..pg.ml.\tCollagen.I.alpha.1..pg.ml.\tComplement.Factor.D.Adipsin..pg.ml.\tDPPIV.CD26..pg.ml.\tFibroblast.Activation.Protein.alpha.FAP..pg.ml.\tGalectin.3..pg.ml.\tGalectin.3BP.MAC.2BP..pg.ml.\tIGFBP.2..pg.ml.\tLumican..pg.ml.\tMyeloperoxidase.MPO..pg.ml.\tVAP.1.AOC3..pg.ml.\tCA15.3.MUC.1..Unit.\tCCL5.RANTES..pg.ml.\tComplement.component.C9..pg.ml.\tEN.RAGE.S100A12..pg.ml.\tM.CSF.R.CD115..pg.ml.\tMMP.2..pg.ml.\tMMP.9..pg.ml.\tMyoglobin..pg.ml.\tLipocalin.2.NGAL..pg.ml.\tTIMP.1..pg.ml.\tCCL19.MIP.3.beta..pg.ml.\tCCL2.MCP.1..pg.ml.\tCCL3.MIP.1.alpha..pg.ml.\tCCL4.MIP.1.beta..pg.ml.\tCD23.Fc.epsilon.RII..pg.ml.\tCD31.PECAM.1..pg.ml.\tCX3CL1.Fractalkine..pg.ml.\tCXCL9.MIG..pg.ml.\tErbB3.Her3..pg.ml.\tFlt.3.Ligand.FLT3L..pg.ml.\tGranzyme.B..pg.ml.\tHE4.WFDC2..pg.ml.\tIFN.gamma..pg.ml.\tIGFBP.4..pg.ml.\tIL.1.beta..pg.ml.\tIL.10..pg.ml.\tIL.12.p40..pg.ml.\tIL.6..pg.ml.\tIL.8..pg.ml.\tKallikrein.6.Neurosin..pg.ml.\tLIGHT.TNFSF14..pg.ml.\tMMP.12..pg.ml.\tMMP.13..pg.ml.\tMMP.3..pg.ml.\tMMP.7..pg.ml.\tMMP.8..pg.ml.\tTGF.alpha..pg.ml.\tTNF.alpha..pg.ml.\tTREM.1..pg.ml.\tTSLP..pg.ml.\tVEGF..pg.ml.\tAlpha.1.Microglobulin..pg.ml.\tAngiopoietin.2..pg.ml.\tBAFF..pg.ml.\tBMP.10..pg.ml.\tContactin.1..pg.ml.\tCXCL11.I.TAC..pg.ml.\tCXCL13.BCA.1..pg.ml.\tCXCL5.ENA.78..pg.ml.\tEndocan.ESAM.1..pg.ml.\tErbB2.Her2..pg.ml.\tFGF.basic.bFGF..pg.ml.\tGalectin.9..pg.ml.\tG.CSF..pg.ml.\tGM.CSF..pg.ml.\tICAM.1.CD54..pg.ml.\tIL.1.alpha.IL.1F1..pg.ml.\tIL.12.p70..pg.ml.\tIL.17.IL.17A..pg.ml.\tIL.1ra.IL.1F3..pg.ml.\tIL.2..pg.ml.\tIL.31..pg.ml.\tIL.4..pg.ml.\tIL.5..pg.ml.\tLeptin.OB..pg.ml.\tMAdCAM.1..pg.ml.\tNT.4..pg.ml.\tSCF.c.kit.Ligand...pg.ml.\tThrombopoietin.Tpo..pg.ml.\tu.Plasminogen.Activator..uPA...pg.ml.\tVCAM.1.CD106..pg.ml.\tCCL22.MDC..pg.ml.\tCCL23.MPIF.1..pg.ml.\tCCL26.Eotaxin.3..pg.ml.\tCD117.c.kit..pg.ml.\tCD163..pg.ml.\tCEACAM.1.CD66a..pg.ml.\tCoagulation.Factor.III.Tissue.Factor..pg.ml.\tFollistatin.related.Gene.Protein.FLRG..pg.ml.\tFurin..pg.ml.\tIFN.beta..pg.ml.\tIGFBP.1..pg.ml.\tIL.11..pg.ml.\tIL.28B.IFN.lamgda.3..pg.ml.\tIL.4R.alpha..pg.ml.\tIL.6R.alpha..pg.ml.\tMCAM.CD146..pg.ml.\tOncostatin.M.OSM...pg.ml.\tOsteoactivin.GPNMB..pg.ml.\tReg3A..pg.ml.\tResistin..pg.ml.\tS100A9..pg.ml.\tSCGF.CLEC11a..pg.ml.\tSerpin.B3.SCCA1..pg.ml.\tTACI.TNFRSF13B..pg.ml.\tThrombomodulin.BDCA.3..pg.ml.\tThrombospondin.2..pg.ml.\tTRAIL.R2.TNFRSF10B..pg.ml.\tTRNACE.TNFSF11.RANK.L..pg.ml.\tADAMTS13..pg.ml.\tAPRIL.TNFSF13..pg.ml.\tCA125.MUC16..Units.\tChitinase.3.like.1..pg.ml.\tCXCL1.GRO.alpha..pg.ml.\tCXCL10.IP.10..pg.ml.\tDkk.1..pg.ml.\tEGF..pg.ml.\tENPP.2.Autotaxin..pg.ml.\tGDF.15..pg.ml.\tGrowth.Hormone..pg.ml.\tHGF..pg.ml.\tIFN.alpha..pg.ml.\tIL.13..pg.ml.\tIL.18.IL.1F4...pg.ml.\tIL.23..pg.ml.\tIL.33..pg.ml.\tM.CSF..pg.ml.\tMICA..pg.ml.\tMIF..pg.ml.\tNeuregulin.1.beta.NRG1.beta.1..pg.ml.\tPentraxin.3.TSG.14..pg.ml.\tSP.D..pg.ml.\tST2.IL.33R..pg.ml.\tuPAR..pg.ml.\tvWF.A2..pg.ml.\tAngiopoietin.1..pg.ml.\tAngiopoietin.like.Protein.3.ANGPTL3..pg.ml.\tCCL11.Eotaxin..pg.ml.\tCCL24.Eotaxin.2.MPIF.2..pg.ml.\tCD40.Ligand.TNFSF5..pg.ml.\tComplement.Componet.C5a..pg.ml.\tCXCL2.Gro.Beta.MIP.2..pg.ml.\tCXCL6.GCP.2..pg.ml.\tFABP4.A.FABP..pg.ml.\tFas.Ligand.TNFSF6..pg.ml.\tGranzyme.A..pg.ml.\tIL.15..pg.ml.\tIL.17E.IL.25..pg.ml.\tIL.3..pg.ml.\tIL.36.beta.IL.1F8..pg.ml.\tIL.7..pg.ml.\tLIF..pg.ml.\tL.Selectin.CD62L..pg.ml.\tMIA..pg.ml.\tNectin.4...pg.ml.\tNephrin..pg.ml.\tOsteopontin.OPN..pg.ml.\tPDGF.AA..pg.ml.\tPD.L1.B7.H1..pg.ml.\tProlactin..pg.ml.\tProprotein.convertase.9.PCSK9..pg.ml.\tP.Selectin.CD62P..pg.ml.\tSyndecan.1.CD138..pg.ml.\tTFF3..pg.ml.\tTRAIL.TNFSF10..pg.ml.\tVEGFR1.Flt.1..pg.ml.\tTGF.beta.1..pg.ml.\tTGF.beta.2..pg.ml.\tBDNF..pg.ml.\tPDFG.DD..pg.ml.\tComplement.C1q..ng.ml.\tSDF.1..pg.ml.\n",
    "# CCO-001-1\t411539661.8 \t626906.9 \t171026177.5\t1448933.7 \t3532220.8\t11603412.1\t850775.5\t17326498.7\t808855.5\t12877186.4\t5841211.6\t7484792.8\t1737816.4\t3123555.2\t27173.3\t7992.3\t16683.4\t9122.3\t44607.5 \t2655\t1792873.4\t96783.8\t7258.2\t5230.5\t853728\t94706.6\t494400.7\t59423.5\t84398.2\t5.8\t14816.7\t106646560.2\t37050.2\t113773.6\t91854.8\t46868.1\t2186.1\t25639\t41856.3\t59.5\t350.8\t244.4 \t318.4 \t1163.6 \t40031.3\t2976.5 \t687.3 \t8149.4\t53.6\t30.4\t3445.9\t6.2 \t75.3 \t<1.34\t4.8 \t375.5 \t5.9 \t9.9 \t1247.4\t0.5 \t<38.97\t700.6 \t18520.3\t3081.7 \t2038\t4.1 \t5.1 \t77.6 \t4.1 \t36.3 \t51600493\t292.3 \t372.1\t60.1 \t2284.8\t89.6 \t54.1 \t207.3 \t318.6\t1196.3 \t13.2 \t27781.5\t71.6 \t9.6 \t139665.3 \t6.5 \t31.7 \t10.0 \t838.3\t13.6 \t<69.26\t47.4 \t5.3 \t2782.2 \t122383.2\t39.0 \t25.3 \t130.0 \t335.4\t1528003.4\t139.6\t188.6\t33.5 \t4133.7 \t283797.7\t42035.3 \t32.7\t9184.9\t2907.5\t16.0 \t35615.9 \t2588.7 \t237.8 \t24.8\t28539\t25967.2\t1163.7\t7929.3\t6701.3\t6523.3\t435.5\t5438.9\t720.5 \t29.9 \t3718.9\t14598.9\t30.1 \t9.3 \t811629.7\t287.5 \t67.9 \t10476.9\t232.1 \t127.5\t629\t43.5 \t23827.3\t535.6\t457.0 \t79.1\t11.7 \t358.6 \t274\t205.3 \t0.7 \t506.5 \t57.8 \t10221.2\t<7.22\t966.8 \t4404.2\t10444.4\t237.7 \t8185.9 \t8180.7\t26597.8\t307.9\t235.9\t1950.9\t151077.9\t709.3\t165.9\t4430.5\t49.2\t138\t25.7\t1975.9\t150.2 \t49.2\t31.5\t57.4 \t296591.8\t455.4 \t413.9\t889.6\t11869.4\t459\t56.8\t20037.7\t70972.4\t24774.7\t2362.2\t1010\t202.6 \t220.2\t16660.9 \t<13.81\t1940.1\t250.7\t43336.2 \t2185.8 \n",
    "### 모든 txt 파일에서 첫 줄 column 이름만 가져오기\n",
    "\n",
    "### 만약 경로가 dir_CYT_01이면 '\\t'을, dir_CYT_02이면 ','를 구분자로 사용\n",
    "\n",
    "def replace_special_chars(col_name):\n",
    "    col_name = re.sub(r'[ \\-\\(\\),/]', '.', col_name)\n",
    "    col_name = re.sub(r'\\.{2,}', '.', col_name)\n",
    "    return col_name\n",
    "\n",
    "def treat_df(df):\n",
    "    df.columns = [replace_special_chars(col) for col in df.columns]\n",
    "    df = df.rename(columns={'Sample.ID.': 'ID'})\n",
    "    ### 만약 ID에 CNC 또는 CNO가 없다면\n",
    "    if not any(df['ID'].str.contains('CNC|CNO')):\n",
    "        df['PNTTM'] = df['ID'].str[-1]\n",
    "        df['PNTTM'] = df['PNTTM'].replace('0', '1')\n",
    "        df['ID'] = df['ID'].str[:-2]\n",
    "    else:\n",
    "        df['PNTTM'] = '1'\n",
    "    df.columns = [col[1:] if col.startswith('.') else col for col in df.columns]\n",
    "    return df\n",
    "\n",
    "def get_cytokine(directory):\n",
    "    all_dfs = []\n",
    "    filenames = os.listdir(directory)\n",
    "    for filename in filenames:\n",
    "        if filename.startswith('COV') and filename.endswith('.txt'):\n",
    "            if directory == dir_CYT_01:\n",
    "                df = pd.read_csv(directory + filename, sep='\\t')\n",
    "                df = treat_df(df)\n",
    "                df['ID'] = 'COV-' + df['ID']\n",
    "            else:\n",
    "                df = pd.read_csv(directory + filename, sep=',')\n",
    "                df = treat_df(df)\n",
    "                if df['ID'].str.contains('CNC|CNO').any():\n",
    "                    df['ID'] = df['ID'].str[:-2]\n",
    "\n",
    "            all_dfs.append(df)\n",
    "    result = pd.concat(all_dfs)\n",
    "    return result\n",
    "\n",
    "cyt_01 = get_cytokine(dir_CYT_01)\n",
    "cyt_02 = get_cytokine(dir_CYT_02)\n",
    "cyt_all = pd.concat([cyt_01, cyt_02])\n",
    "\n",
    "print(cyt_all.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1159 \n",
      " \n",
      " ID                                     0\n",
      "alpha.2.Macroglobulin.pg.ml.           0\n",
      "beta.2.Microglobulin.pg.ml.          314\n",
      "Fetuin.A.AHSG.pg.ml.                   0\n",
      "MBL.pg.ml.                             0\n",
      "                                    ... \n",
      "CCL14.HCC.1.HCC.3.pg.ml.             845\n",
      "Complement.Component.C2.pg.ml.       845\n",
      "Serpin.C1.Antithrombin.III.pg.ml.    845\n",
      "beta2.Microglobulin.pg.ml.           845\n",
      "CXCL12.pg.ml.                        845\n",
      "Length: 227, dtype: int64\n",
      "\n",
      " 554\n",
      "['1' '2' '3' '4' '5' '6' '7']\n"
     ]
    }
   ],
   "source": [
    "### 전체 row 개수, cyt_all에서 NaN인 row 개수 확인\n",
    "\n",
    "print(cyt_all.shape[0], '\\n', '\\n', cyt_all.isnull().sum())\n",
    "\n",
    "### unique한 ID 개수 확인\n",
    "\n",
    "print('\\n',cyt_all['ID'].unique().shape[0])\n",
    "print(cyt_all['PNTTM'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "261\n",
      "\n",
      " 483 \n",
      " \n",
      " ID                                     0\n",
      "alpha.2.Macroglobulin.pg.ml.           0\n",
      "beta.2.Microglobulin.pg.ml.          117\n",
      "Fetuin.A.AHSG.pg.ml.                   0\n",
      "MBL.pg.ml.                             0\n",
      "                                    ... \n",
      "CCL14.HCC.1.HCC.3.pg.ml.             366\n",
      "Complement.Component.C2.pg.ml.       366\n",
      "Serpin.C1.Antithrombin.III.pg.ml.    366\n",
      "beta2.Microglobulin.pg.ml.           366\n",
      "CXCL12.pg.ml.                        366\n",
      "Length: 227, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "### cyt_all에서 ID가 max_list에 있는 경우만 cyt_adapt에 저장\n",
    "\n",
    "cyt_adapt = cyt_all[cyt_all['ID'].isin(max_list)]\n",
    "\n",
    "print(cyt_adapt['ID'].unique().shape[0])\n",
    "\n",
    "# 누락 된 4개 ID는 ['COV-CNC-047', 'COV-CNC-057', 'COV-CNC-058', 'COV-CNC-060']\n",
    "\n",
    "### Nan 값이 있는 column 확인 (ID unique기준)\n",
    "\n",
    "print('\\n', cyt_adapt.shape[0], '\\n', '\\n', cyt_adapt.isnull().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "beta.2.Microglobulin.pg.ml.\n",
      "Aldehyd.Dehydrogenase.1.A1.pg.ml.\n",
      "Coagulation.Factor.XIV.Protein.C.pg.ml.\n",
      "u.Plasminogen.Activator.uPA.pg.ml.\n",
      "Coagulation.Factor.III.Tissue.Factor.pg.ml.\n",
      "TRNACE.TNFSF11.RANK.L.pg.ml.\n",
      "CXCL1.GRO.alpha.pg.ml.\n",
      "Neuregulin.1.beta.NRG1.beta.1.pg.ml.\n",
      "Complement.Componet.C5a.pg.ml.\n",
      "Proprotein.convertase.9.PCSK9.pg.ml.\n",
      "TRANCE.TNFSF11.RANK.L.pg.ml.\n",
      "Neuregulin.1.beta.1.NRG1.beta.1.pg.ml.\n",
      "Proprotein.Convertase.9.PCSK9.pg.ml.\n",
      "Coagulation.Factor.II.Tissue.Factor.pg.ml.\n",
      "Complement.Component.C5a.pg.ml.\n",
      "CXCL1.GRO.alpha.KC.CINC.1.pg.ml.\n",
      "u.Plasminogen.Activator.uPA.Urokinase.pg.ml.\n",
      "Complement.Component.C9.pg.ml.\n",
      "Aldehyde.Dehydrogenase.1.A1.ALDH1A1.pg.ml.\n",
      "Complement.Component.C2.pg.ml.\n",
      "beta2.Microglobulin.pg.ml.\n",
      "21\n"
     ]
    }
   ],
   "source": [
    "### cyt_adapt에서 맨 뒤의 7자리를 제외하고 n글자 이상 동일한 문자열이 있는 column 확인\n",
    "\n",
    "\n",
    "\n",
    "def get_similar_cols(df, length):\n",
    "    similar_cols = []\n",
    "    \n",
    "    def get_substrings(s):\n",
    "        if len(s) < length:\n",
    "            return []\n",
    "        return [s[i:i+length] for i in range(len(s) - length + 1)]\n",
    "    \n",
    "    for col in df.columns:\n",
    "        if len(col) > length:\n",
    "            col_substrings = set(get_substrings(col[:-7]))\n",
    "            for col2 in df.columns:\n",
    "                if col != col2 and len(col2) > length:\n",
    "                    col2_substrings = set(get_substrings(col2[:-7]))\n",
    "                    if col_substrings.intersection(col2_substrings):\n",
    "                        similar_cols.append(col)\n",
    "                        break\n",
    "    return similar_cols\n",
    "\n",
    "similar_cols = get_similar_cols(cyt_adapt,15)\n",
    "for i in similar_cols:\n",
    "    print(i, sep='\\n')\n",
    "print(len(similar_cols))\n",
    "        \n",
    "for i in similar_cols:\n",
    "    if not i.endswith('pg.ml.'):\n",
    "        print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
